### ISSUE TITLE
trainer.summarize_test_progress() and eval_test_images() return two different accuracies.

### ISSUE THREAD

**Issue by MaximeLeclerc** [Microsoft/CNTK#2874]
Hi,

I'm comparing trainer.summarize_test_progress() and eval_test_images() using https://github.com/Microsoft/CNTK/blob/master/Examples/Image/TransferLearning/TransferLearning.py .

Unfortunately, the two functions return two different accuracies.

Here is the code I added in order to call trainer.summarize_test_progress():

```
# Create the test minibatch source and input variables                                                                                                                    
test_size = sum(1 for line in open(test_map_file))                                                                                                                        
minibatch_source_test = create_mb_source(test_map_file, image_width, image_height, num_channels, num_classes)

tensorboard_writer = TensorBoardProgressWriter(freq=10, log_dir=train_dir_name, model=tl_model)                                                                           
trainer = Trainer(tl_model, (ce, pe), learner, [progress_printer, tensorboard_writer])

          test_count = 0                                                                                                                                                        
          while test_count < test_size:  # loop over minibatches in the test epoch                                                                                              
              data_test = minibatch_source_test.next_minibatch(min(mb_size, test_size-test_count), input_map=input_map)                                                         
              trainer.test_minibatch(data_test)                                # test model with it                                                                             
              test_count += data_test[label_input].num_samples                 # count test samples processed so far                                                            
                                                                                                                                                                                
          trainer.summarize_test_progress()
```

Here are the accuracies I get:

```
Finished Epoch[15 of 20]: [Training] loss = 0.001291 * 6149, metric = 0.02% * 6149 11.828s (519.9 samples/s);
Finished Evaluation [15]: Minibatch[1-21]: metric = 5.98% * 1020;
Processed 5000 samples
Finished Epoch[16 of 20]: [Training] loss = 0.001258 * 6149, metric = 0.02% * 6149 11.677s (526.6 samples/s);
Finished Evaluation [16]: Minibatch[1-21]: metric = 6.18% * 1020;
Processed 5000 samples
Finished Epoch[17 of 20]: [Training] loss = 0.001432 * 6149, metric = 0.03% * 6149 11.839s (519.4 samples/s);
Finished Evaluation [17]: Minibatch[1-21]: metric = 6.08% * 1020;
Processed 5000 samples
Finished Epoch[18 of 20]: [Training] loss = 0.001343 * 6149, metric = 0.02% * 6149 12.015s (511.8 samples/s);
Finished Evaluation [18]: Minibatch[1-21]: metric = 6.08% * 1020;
Processed 5000 samples
Finished Epoch[19 of 20]: [Training] loss = 0.001367 * 6149, metric = 0.02% * 6149 11.783s (521.9 samples/s);
Finished Evaluation [19]: Minibatch[1-21]: metric = 6.18% * 1020;
Processed 5000 samples
Finished Epoch[20 of 20]: [Training] loss = 0.001320 * 6149, metric = 0.02% * 6149 11.864s (518.3 samples/s);
Finished Evaluation [20]: Minibatch[1-21]: metric = 5.98% * 1020;
Stored trained model at /home/maxime/CNTK-Samples-2-3-1/Examples/Image/TransferLearning/resnet_20180123_163115/TransferLearning.model
Evaluating model output node 'prediction' for 1020 images.
Processed 100 samples (0.86 correct)
Processed 200 samples (0.91 correct)
Processed 300 samples (0.93 correct)
Processed 400 samples (0.925 correct)
Processed 500 samples (0.928 correct)
Processed 600 samples (0.9366666666666666 correct)
Processed 700 samples (0.9428571428571428 correct)
Processed 800 samples (0.9475 correct)
Processed 900 samples (0.9477777777777778 correct)
Processed 1000 samples (0.944 correct)
964 out of 1020 predictions were correct 0.9450980392156862.
```
Unfortunately, 1-0.9450980392156862=0.05490196078 which is different from 5.98%.

Any idea what the problem could be?

Thanks a lot and have a great day,   Maxime

**Comment by KeDengMS** [comment]
Are you referring to this log?

Finished Evaluation [20]: Minibatch[1-21]: metric = 5.98% * 1020;

This is the metric for just Epoch 20, while 0.945 is for all samples.

**Comment by MaximeLeclerc** [comment]
Hi @KeDengMS 

Our goal is to log the test accuracy after each sequence of training is completed. However since the two methods give different results we're hesitant to use either method. In addition, `trainer.test_minibatch()` is much faster than `eval_test_images()` and would advantageous for our large dataset that we plan to train once we get the Flowers dataset working.

I'm not sure I see a difference in the network and samples the metrics are computed with?

- The first metric is computed with the final trained network just before `return tl_model` in `train_model()`. It takes as a second input the 1020 test files listed in `1k_img_map.txt`:

`Finished Evaluation [20]: Minibatch[1-21]: metric = 5.98% * 1020;`

- The second metric is computed with the final trained network just after `return tl_model` in `train_model()`. It takes as a second input the 1020 test files listed in `1k_img_map.txt`:

`964 out of 1020 predictions were correct 0.9450980392156862.`

Shouldn't the computed accuracies be equal?

We also did a test where we called `trainer.summarize_test_progress()` for each minibatch just before `return tl_model` to diagnose further what might be going on. Here are the results:

```
Finished Evaluation [419]: Minibatch[1-1]: metric = 2.00% * 50;
Finished Evaluation [420]: Minibatch[1-1]: metric = 8.00% * 50;
Finished Evaluation [421]: Minibatch[1-1]: metric = 4.00% * 50;
Finished Evaluation [422]: Minibatch[1-1]: metric = 8.00% * 50;
Finished Evaluation [423]: Minibatch[1-1]: metric = 10.00% * 50;
Finished Evaluation [424]: Minibatch[1-1]: metric = 2.00% * 50;
Finished Evaluation [425]: Minibatch[1-1]: metric = 12.00% * 50;
Finished Evaluation [426]: Minibatch[1-1]: metric = 10.00% * 50;
Finished Evaluation [427]: Minibatch[1-1]: metric = 2.00% * 50;
Finished Evaluation [428]: Minibatch[1-1]: metric = 8.00% * 50;
Finished Evaluation [429]: Minibatch[1-1]: metric = 8.00% * 50;
Finished Evaluation [430]: Minibatch[1-1]: metric = 6.00% * 50;
Finished Evaluation [431]: Minibatch[1-1]: metric = 0.00% * 50;
Finished Evaluation [432]: Minibatch[1-1]: metric = 10.00% * 50;
Finished Evaluation [433]: Minibatch[1-1]: metric = 8.00% * 50;
Finished Evaluation [434]: Minibatch[1-1]: metric = 10.00% * 50;
Finished Evaluation [435]: Minibatch[1-1]: metric = 10.00% * 50;
Finished Evaluation [436]: Minibatch[1-1]: metric = 6.00% * 50;
Finished Evaluation [437]: Minibatch[1-1]: metric = 6.00% * 50;
Finished Evaluation [438]: Minibatch[1-1]: metric = 2.00% * 50;
Finished Evaluation [439]: Minibatch[1-1]: metric = 5.00% * 20;

[...]

958 out of 1020 predictions were correct 0.9392156862745098.
```

This indicates that `67/1020 = 0.0656862745` samples were not classified correctly according to `trainer.test_minibatch()` just before `return tl_model`.

However, `62/1020 = 1-0.9392156862745098 = 0.0607843137` samples were not classified correctly according to `eval_test_images()` just after `return tl_model`.

Maybe there's something going on that we missed?

Thanks a lot and have a great day,   Maxime

**Comment by MaximeLeclerc** [comment]
Hi @KeDengMS 

Note: We also tried calling `trainer.test_minibatch()` after `return tl_model`. It gives the same result that we get when we call `trainer.test_minibatch()` before `return tl_model`. This result is different from calling `eval_test_images()` just after `return tl_model`.

Thanks a lot and have a great day,   Maxime

**Comment by KeDengMS** [comment]
SummarizeTestProgress() shows aggregated test eval criterion value, which is over all samples in one epoch. eval_test_images() in TransferLearning.py runs with a batch_size=1, and also output eval results to text file which adds CPU/GPU sync and I/O cost. I think that's why it's slower than trainer.test_minibatch().

Besides, please don't get confused with loss/error in Training vs. in test. Here's a simple explanation:
```
x = C.input_variable((5,))
y = C.input_variable((5,))
s = C.cross_entropy_with_softmax(x,y) + C.parameter(())
t = C.Trainer(s, (s,s), C.sgd(s.parameters, lr=1))
feed1 = {x:np.random.rand(10,5), y:np.random.rand(10,5)}
t.test_minibatch(feed1)
t.train_minibatch(feed1)
t.previous_minibatch_loss_average
t.test_minibatch(feed1)
```

You'll notice that training loss is evaluated before model update, and test loss after model update would be different.

**Comment by MaximeLeclerc** [comment]
Hi @KeDengMS 

Right. In your example, would you expect the second t.test_minibatch(feed1) on the last line to return the same accuracy than eval_test_images()? Because it does not.

Thanks a lot and have a great day,   Maxime

**Comment by KeDengMS** [comment]
No it does not, because the model has changed in t.train_minibatch(feed1). If you run `np.sum(s.eval(feed1))` you'll get the same output as the second test_minibatch(feed1)
